### PR TITLE
DIRKRB-692 - populates SGT clientPrincipal and fixes storeTicket method 

### PR DESCRIPTION
**Pull request by ftarlao**
For details refer to the issue:
https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

### CODE REVIEW

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java [hunk 1]
```diff
@@ -271,15 +271,23 @@ public void storeTicket(TgtTicket tgtTicket,
...
         }
         if (ccacheFile.exists() && ccacheFile.canWrite()) {
-            CredentialCache cCache = new CredentialCache();
+            
             try {
-                cCache.load(ccacheFile);
-                cCache.addCredential(new Credential(sgtTicket, sgtTicket.getClientPrincipal()));
-                cCache.setPrimaryPrincipal(sgtTicket.getClientPrincipal());
+                CredentialCache cCache;
+                if(!isFreshNew){
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: Please put spaces before "(" and after ")".

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java [hunk 2]
```diff
@@ -271,15 +271,23 @@ public void storeTicket(TgtTicket tgtTicket,
...
+                    //Remind, contructor sets the cCache client principal from the sgtTicket one
+                    cCache = new CredentialCache(sgtTicket);
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: Why not "new CredentialCache()" here as the code was before?

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java [hunk 3]
```diff
@@ -153,13 +153,14 @@ public SgtTicket requestSgt(KOptions requestOptions) throws KrbException {
...
 
+        PrincipalName clientPrincipalName = null;
         if (tgtTicket != null) {
             String sourceRealm = tgtTicket.getRealm();
             String destRealm = serverPrincipalName.getRealm();
             if (!sourceRealm.equals(destRealm)) {
                 KrbConfig krbConfig = krbSetting.getKrbConfig();
                 LinkedList<String> capath = krbConfig.getCapath(sourceRealm, destRealm);
-                PrincipalName clientPrincipalName = tgtTicket.getClientPrincipal();
+                clientPrincipalName = tgtTicket.getClientPrincipal();
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: This line doesn't apply if the source realm is equal to the dest realm, so you could move it above the if statement.

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java [hunk 4]
```diff
@@ -170,11 +171,19 @@ public SgtTicket requestSgt(KOptions requestOptions) throws KrbException {
                     tgsRequest = new TgsRequestWithTgt(context, sgtTicket);
                 }
             }
+        } else {
+            //This code is for the no-tgt case but should be double-checked
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: Is this part required for your use-case? If not I think it could be left out.

**Review by coheigea — CHANGES_REQUESTED** [review]

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java (hunk 3 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: right, I'll fix

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java (hunk 4 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: I think that trying to support a specific 'manage use-case' is wrong (IMHO), I always try to manage all the use-cases :-) . I have a question, is there a way to request an sgt without tgt? (e.g., OTP, user/pass) and in that case, where is the principal name stored?  I have speculated (in the KOptions) CLIENT_PRINCIPAL. Or is the USE_TGT the only accepted case for requestSGT?

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java (hunk 1 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: I'll fix

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java (hunk 2 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: right..

Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: I should also remove, in the same block, the cCache.setPrimaryPrincipal(sgtTicket.getClientPrincipal()); because adding an SGT should not change the primary client principal.. I think

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java (hunk 1 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: I should also improve the condition: isFreshNew = !ccacheFile.exists() || (ccacheFile.length() == 0); in order to manage corner cases (opening a file have caused me problems in the past about automatic file creation... of empty files, don't know if this is OS related).

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java (hunk 4 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: At least, I'll leave an empty 'else' block with a reminder .. for other possible use cases...

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java [hunk 5]
```diff
@@ -271,15 +271,22 @@ public void storeTicket(TgtTicket tgtTicket,
...
+        if (isFreshNew) {
             createCacheFile(ccacheFile);
         }
         if (ccacheFile.exists() && ccacheFile.canWrite()) {
-            CredentialCache cCache = new CredentialCache();
+            
             try {
-                cCache.load(ccacheFile);
-                cCache.addCredential(new Credential(sgtTicket, sgtTicket.getClientPrincipal()));
-                cCache.setPrimaryPrincipal(sgtTicket.getClientPrincipal());
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: Why is this line removed?

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java (hunk 4 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: I also improved in another comments the clientPrincipal "null" case.. please refer to commit comments for details

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java [hunk 6]
```diff
@@ -170,11 +171,22 @@ public SgtTicket requestSgt(KOptions requestOptions) throws KrbException {
...
+                clientPrincipalName = new PrincipalName(clientPrincipalString);
+            }
         }
-
+        
         tgsRequest.setServerPrincipal(serverPrincipalName);
         tgsRequest.setRequestOptions(requestOptions);
-        return doRequestSgt(tgsRequest);
+        SgtTicket sgtTicket = doRequestSgt(tgsRequest);
+        if(clientPrincipalName!=null){
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: Please put spaces around brackets here.

**Review by coheigea — CHANGES_REQUESTED** [review]

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java (hunk 6 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java]: Gosh, I hate formatting :-) .. I'll fix this

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java (hunk 5 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: When ccache exists and is correctly loaded, the cCache primaryPrincipal field should be already populated. Furthermore, I think that adding an SGT should not change the cCache primary principal (when present). I think that, setting the primary principal, is only needed when  fresh new cache file is created.

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java [hunk 7]
```diff
@@ -271,15 +271,22 @@ public void storeTicket(TgtTicket tgtTicket,
      */
     public void storeTicket(SgtTicket sgtTicket, File ccacheFile) throws KrbException {
         LOG.info("Storing the sgt to the credential cache file.");
-        if (!ccacheFile.exists()) {
+        boolean isFreshNew = !ccacheFile.exists() || (ccacheFile.length() == 0);
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: Please change "isFreshNew" to "createCache".

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java [hunk 8]
```diff
@@ -271,15 +271,22 @@ public void storeTicket(TgtTicket tgtTicket,
...
+        if (isFreshNew) {
             createCacheFile(ccacheFile);
         }
         if (ccacheFile.exists() && ccacheFile.canWrite()) {
-            CredentialCache cCache = new CredentialCache();
+
```
Comment (coheigea) [inline kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: There is some trailing whitespace here, and in a few other places as well that could be removed.

**Review by coheigea — CHANGES_REQUESTED** [review]

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java (hunk 7 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: done

File: kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java (hunk 8 above)
Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: I'm going to commit the changes, removed the blank rows.. I left only the blank rows that were in the original code.

Comment (ftarlao) [reply kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java]: Ok, I totally misunderstood. Trailing spaces are not empty lines! :-( English is not my natural language and sometimes I fail. I added a new commit to the pull request, deleted the trailing spaces and re-added few blank lines around if/else blocks.. that.. I think ... improve readability.
Regards

**Comment by coheigea** [comment]
There is still a lot of trailing whitespace that I can see when I apply the patch. For example, line 279 of KrbClientBase.java.

**Comment by ftarlao** [comment]
I'm sorry, I'm not so good about formatting.. I think you have some
automated check for this and that you havee strict rules.
I have tried now with a search regex.. hope to have solved all
trailing-space-issues
